### PR TITLE
DYN-4231 Fast Path for sweep during FullGC

### PR DESCRIPTION
**Pull request by saintentropy**
### Purpose

https://jira.autodesk.com/browse/DYN-4231

The purpose of this PR is to optimize the Sweep pass for the FullGC pass that occurs after any change to the graph that causes re-execution.   Specifically this optimization does the rolling:

*  introduces a specific Call method to the Executive (`CallDispose`) tailored specifically for dispatching Dispose calls vs allowing these calls from the GC to pass through the generic `Callr`.  Additionally it introduces a specific Dispatch method to the Callsite ('DispatchDispose`) also tailored to Dispatch Dispose calls vs `Dispatch`.  In both case the normal overhead of the `Callr` and `Dispatch` is much higher and .NET memory/GC intensive than the actual `Dispose` calls.
* Adds caching based on object type where possible as the GC is often collecting similar types repeatedly

In testing a graph that had ~150000 items to collect in the sweep pass the time was reduced from 2s to .5s and the .NET memory allocation from 800mb to 16mb.

Todo is to validate with testing include DS object dispose.

### Declarations

Check these if you believe they are true

- [x] The codebase is in a better state after this PR
- [x] Is documented according to the [standards](https://github.com/DynamoDS/Dynamo/wiki/Coding-Standards)
- [ ] The level of testing this PR includes is appropriate
- [ ] User facing strings, if any, are extracted into `*.resx` files
- [x] All tests pass using the self-service CI.
- [ ] Snapshot of UI changes, if any.
- [ ] Changes to the API follow [Semantic Versioning](https://github.com/DynamoDS/Dynamo/wiki/Dynamo-Versions) and are documented in the [API Changes](https://github.com/DynamoDS/Dynamo/wiki/API-Changes) document.
- [ ] This PR modifies some build requirements and the readme is updated

### Reviewers

TBD

### FYIs

@jasonstratton @QilongTang

### CODE REVIEW

File: src/Engine/ProtoCore/DSASM/Executive.cs [hunk 1]
```diff
@@ -817,6 +817,36 @@ private List<StackValue> PopArgumentsFromStack(int argumentCount)
             return sv;
         }
 
+        // cached values for repeated calls of the same object type and dispose function.
+        private int previousClassIndex;
+        private string previousProcedureName;
+        private CallSite callsite;
+
+        public StackValue CallDipose(ProcedureNode fNode,
```
Comment (sm6srw) [inline src/Engine/ProtoCore/DSASM/Executive.cs:825]: Spelling `CallDispose`

File: src/Engine/ProtoCore/DSASM/Executive.cs [hunk 2]
```diff
@@ -817,6 +817,36 @@ private List<StackValue> PopArgumentsFromStack(int argumentCount)
...
+            {
+                previousClassIndex = classIndex;
+                previousProcedureName = fNode.Name;
+                callsite = new CallSite(classIndex, fNode.Name, exe.FunctionTable, runtimeCore.Options.ExecutionMode);
+            }
+
+            Validity.Assert(null != callsite);
+
+            StackValue sv = StackValue.Null;
+            sv = callsite.DispatchDispose(svThisPtr, runtimeCore);
```
Comment (aparajit-pratap) [inline src/Engine/ProtoCore/DSASM/Executive.cs:845]: We can simply combine these statements into one:
```
StackValue sv = callsite.DispatchDispose(svThisPtr, runtimeCore);
```

File: src/Engine/ProtoCore/DSASM/Heap.cs [hunk 3]
```diff
@@ -624,40 +624,45 @@ private bool TryFindFreeIndex(out int index)
             }
         }
 
+        //cache ClassIndex and ProcdureNode for repeated calls of the same type object.
+        private int previousClassIndex;
+        private ProcedureNode pn;
```
Comment (aparajit-pratap) [inline src/Engine/ProtoCore/DSASM/Heap.cs:629]: I think we should rename this to something more descriptive for clarity.

File: src/Engine/ProtoCore/DSASM/Heap.cs [hunk 4]
```diff
@@ -624,40 +624,45 @@ private bool TryFindFreeIndex(out int index)
...
                 }
             }
 
             if (pn != null)
             {
-                // TODO Jun/Jiong: Use build pointer utilities 
-                exe.rmem.Push(StackValue.BuildArrayDimension(0));
-                exe.rmem.Push(StackValue.BuildPointer(svPtr.Pointer, svPtr.metaData));
-                exe.rmem.Push(StackValue.BuildInt(1));
-
```
Comment (aparajit-pratap) [inline src/Engine/ProtoCore/DSASM/Heap.cs:653]: Why have these lines been removed?

File: src/Engine/ProtoCore/Lang/CallSite.cs [hunk 5]
```diff
@@ -1528,6 +1528,27 @@ private FunctionGroup FirstFunctionGroupInInheritanceChain(RuntimeCore runtimeCo
             return ret;
         }
 
+        //Pre-initialize array for repeated calls.  The StackValue is inserted vs making a new array for every call.
+        private static List<StackValue> disposeArguments = new List<StackValue>() { StackValue.Null };
+        //Cache final function endpoint for repeated calls;
+        private FunctionEndPoint finalFep;
+
+        public StackValue DispatchDispose(
```
Comment (aparajit-pratap) [inline src/Engine/ProtoCore/Lang/CallSite.cs:1536]: Why is this public?

File: src/Engine/ProtoCore/DSASM/Executive.cs (hunk 1 above)
Comment (aparajit-pratap) [reply src/Engine/ProtoCore/DSASM/Executive.cs:825]: +1

File: src/Engine/ProtoCore/Lang/CallSite.cs [hunk 6]
```diff
@@ -1528,6 +1528,27 @@ private FunctionGroup FirstFunctionGroupInInheritanceChain(RuntimeCore runtimeCo
...
+            StackValue stackValue, RuntimeCore runtimeCore)
```
Comment (aparajit-pratap) [inline src/Engine/ProtoCore/Lang/CallSite.cs:1537]: Can we have this line on the same line as 1536?

File: src/Engine/ProtoCore/Lang/CallSite.cs [hunk 7]
```diff
@@ -1528,6 +1528,27 @@ private FunctionGroup FirstFunctionGroupInInheritanceChain(RuntimeCore runtimeCo
...
+        {
+            if (finalFep == null)
+            {
+                var funcGroup = FirstFunctionGroupInInheritanceChain(runtimeCore, classScope);
+                finalFep = funcGroup.FunctionEndPoints[0];
```
Comment (aparajit-pratap) [inline src/Engine/ProtoCore/Lang/CallSite.cs:1542]: Why is the Dispose method always the first fep?

File: src/Engine/ProtoCore/Lang/FFIFunctionEndPoint.cs [hunk 8]
```diff
@@ -84,9 +84,7 @@ private void Init(RuntimeCore runtimeCore)
         }
 
         public override StackValue Execute(Runtime.Context c, List<StackValue> formalParameters, StackFrame stackFrame, RuntimeCore runtimeCore)
-        {   
-            StackValue svThisPtr = stackFrame.ThisPtr;
```
Comment (aparajit-pratap) [inline src/Engine/ProtoCore/Lang/FFIFunctionEndPoint.cs:88]: Was the stackFrame null in some cases? If so, was this crashing before, is this why you added the null check in line 98?

File: src/Engine/ProtoCore/DSASM/Heap.cs [hunk 9]
```diff
@@ -624,40 +624,45 @@ private bool TryFindFreeIndex(out int index)
...
-                
                 ++exe.RuntimeCore.FunctionCallDepth;
 
                 // TODO: Need to move IsExplicitCall to DebugProps and come up with a more elegant solution for this
                 // fix for IDE-963 - pratapa
                 bool explicitCall = exe.IsExplicitCall;
                 bool tempFlag = explicitCall;
-                exe.Callr(pn.RuntimeIndex, pn.ID, classIndex, ref explicitCall);
+                //exe.Callr(pn.RuntimeIndex, pn.ID, classIndex, ref explicitCall);
+                exe.CallDipose(pn, svPtr, classIndex, ref explicitCall);
```
Comment (aparajit-pratap) [inline src/Engine/ProtoCore/DSASM/Heap.cs:665]: `explicitCall` is not being touched inside of `CallDispose()`? Can we check this? As far as I remember this had some implications in the case of Dispose calls. Is it no longer relevant?

File: src/Engine/ProtoCore/Lang/CallSite.cs (hunk 7 above)
Comment (aparajit-pratap) [reply src/Engine/ProtoCore/Lang/CallSite.cs:1542]: Has this been tested with DesignScript objects? Is the Dispose method defined on a DS class called after these changes?

**Review by sm6srw — COMMENTED** [review]
Many thanks @aparajit-pratap for your insights. I have been trying to break this in various ways, unsuccessfully so far. It works well. In many cases where it currently takes 10-15 times longer to dispose objects than it takes to create them it now takes about the same time to create and dispose.  I am also seeing about 20% less memory being used.

File: src/Engine/ProtoCore/DSASM/Executive.cs (hunk 1 above)
Comment (saintentropy) [reply src/Engine/ProtoCore/DSASM/Executive.cs]: Spelling fixed and set to internal

File: src/Engine/ProtoCore/DSASM/Executive.cs (hunk 2 above)
Comment (saintentropy) [reply src/Engine/ProtoCore/DSASM/Executive.cs]: Good call... fixed

File: src/Engine/ProtoCore/DSASM/Heap.cs (hunk 3 above)
Comment (saintentropy) [reply src/Engine/ProtoCore/DSASM/Heap.cs]: I renamed it to `disposeProcedureNode`

File: src/Engine/ProtoCore/DSASM/Heap.cs [hunk 10]
```diff
@@ -624,40 +624,45 @@ private bool TryFindFreeIndex(out int index)
...
-                exe.rmem.Push(StackValue.BuildInt(1));
-                
                 ++exe.RuntimeCore.FunctionCallDepth;
 
                 // TODO: Need to move IsExplicitCall to DebugProps and come up with a more elegant solution for this
                 // fix for IDE-963 - pratapa
                 bool explicitCall = exe.IsExplicitCall;
                 bool tempFlag = explicitCall;
-                exe.Callr(pn.RuntimeIndex, pn.ID, classIndex, ref explicitCall);
+                //exe.Callr(pn.RuntimeIndex, pn.ID, classIndex, ref explicitCall);
```
[the remark this answers is not in the document]
Comment (saintentropy) [reply src/Engine/ProtoCore/DSASM/Heap.cs]: deleted

File: src/Engine/ProtoCore/DSASM/Heap.cs (hunk 4 above)
Comment (saintentropy) [reply src/Engine/ProtoCore/DSASM/Heap.cs]: This was not necessary with the new route and would be a memory leak.  Also FYI in Dynamo today it does seem to have a memory leak regardless.  We always add 3 items but only ever Pop two.

File: src/Engine/ProtoCore/Lang/CallSite.cs (hunk 5 above)
Comment (saintentropy) [reply src/Engine/ProtoCore/Lang/CallSite.cs]: moved to internal

File: src/Engine/ProtoCore/Lang/CallSite.cs (hunk 6 above)
Comment (saintentropy) [reply src/Engine/ProtoCore/Lang/CallSite.cs]: Fixed

File: src/Engine/ProtoCore/Lang/CallSite.cs (hunk 7 above)
Comment (saintentropy) [reply src/Engine/ProtoCore/Lang/CallSite.cs:1542]: I pulled some test back out of the past and it did fail.  Not sure how much this case needs to be covered but I updated the PR to handle legacy DS object dispose with the original callr

File: src/Engine/ProtoCore/Lang/FFIFunctionEndPoint.cs (hunk 8 above)
Comment (saintentropy) [reply src/Engine/ProtoCore/Lang/FFIFunctionEndPoint.cs:88]: This check is new.  The `stackFrame` is not passed here when being invoked from the `DisposeDispatch`.  This is now a check for that case so it skips the processing necessary for a generic call and jumps to Execute.

File: src/Engine/ProtoCore/DSASM/Heap.cs (hunk 9 above)
Comment (saintentropy) [reply src/Engine/ProtoCore/DSASM/Heap.cs]: cleaned up

**Comment by sm6srw** [comment]
@saintentropy You have 4 test failures now.

File: src/Engine/ProtoCore/Lang/CallSite.cs [hunk 11]
```diff
@@ -1528,6 +1528,25 @@ private FunctionGroup FirstFunctionGroupInInheritanceChain(RuntimeCore runtimeCo
...
+            }
+            
+            disposeArguments[0] = stackValue;
+
+            //EXECUTE
+            return finalFep.Execute(null, disposeArguments, null, runtimeCore);
```
Comment (aparajit-pratap) [inline src/Engine/ProtoCore/Lang/CallSite.cs:1548]: Why is the context (first arg) passed as null?

Comment (saintentropy) [reply src/Engine/ProtoCore/Lang/CallSite.cs:1548]: Only because for this execution path it was not referenced.

**Review by aparajit-pratap — APPROVED** [review]

**Review by sm6srw — APPROVED** [review]

**Pull request merged**